# Data Exploration with Python

This exercise mainly follows https://www.districtdatalabs.com/data-exploration-with-python-1 to learn: 
- data exploration methodologies,
- libraries, API
- python code.

## Setup a local operation on a Macbook.

```
➜  dataexplore git:(master) ✗ pwd
/Users/rkuo/code/machinelearning/dataexplore
```

- Create a project directory, called vehicles 
- Start a tensorflow container.
![start a tensorflow container](https://www.evernote.com/l/AS6WmlGLtnZN67v4abl2O-Lkv4yWTkChhPYB/image.png)

## Open a new python3 notebook and rename it to dataexplore-vehicles

In [5]:
# double check the location in container, we need to map it later to save it or export it.
!pwd

/home/jovyan/work/vehicles


## Create a data directory, Download file from source url, and unzip it.

In [2]:
import os
import zipfile
import requests

def download_data(url, name, path='data'):
   . if not os.path.exists(path):
        os.mkdir(path)

    response = requests.get(url)
    with open(os.path.join(path, name), 'wb') as f:
        f.write(response.content)

    z = zipfile.ZipFile(os.path.join(path, 'vehicles.zip'))
    z.extractall(path)

VEHICLES = 'http://bit.ly/ddl-cars'

download_data(VEHICLES, 'vehicles.zip')

 ## Load it into a pandas data frame.

In [6]:
import pandas as pd

path = 'data'
vehicles = pd.read_csv(os.path.join(path, 'vehicles.csv'))

/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


The warning message "/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types. Specify dtype option on import or set low_memory=False. interactivity=interactivity, compiler=compiler, result=result)" is wiered, data only reach to col-CE, which is (26 x 2 + 5 = 57) columns, but 70 = 2 x 26 + 18. Since it is a warning, ignore.

In [7]:
# define an array of selected columns
select_columns = ['make', 'model', 'year', 'displ', 'cylinders', 'trany', 'drive', 'VClass','fuelType','barrels08', 'city08', 'highway08', 'comb08', 'co2TailpipeGpm', 'fuelCost08']
vehicles = vehicles[select_columns][vehicles.year <= 2016].drop_duplicates().dropna()
vehicles = vehicles.sort_values(['make', 'model', 'year'])